sql/schemachanger: Implement CREATE/DROP POLICY in the DSC

spilchen opened the pull request:
Prior to this commit, attempts to create or drop a policy succeeded but did not write any metadata to the table descriptor. This commit introduces the first step in persisting policy metadata by storing the policy name. Future PRs will handle storing additional policy details.

The policy name is now stored in a new field within the table descriptor, and each policy is assigned a unique ID specific to the table. This commit also adds elements to the DSC to support adding and removing policy information in the descriptor. Support in the legacy schema changer has been intentionally left out.

Epic: CRDB-11724
Informs #136696
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138651)
<!-- Reviewable:end -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138651)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138651#-OG6l0j50E0WUaxWvIXN:-OG6l0j6APqHmw03PFPY:bvhp1ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d976034f574b2c23e9a45523ef1c5b0421de34/pkg/sql/logictest/testdata/logic_test/row_level_security#L13)):*
> ```
> subtest legacy_fail
> 
> let $use_decl_sc
> ```

This is nice, I didn't know we could set variables like this in our tests.  

But what is the advantage to use this variable?
Wouldn't it be clearer to always use `SET use_declarative_schema_changer = 'on';` , `SET use_declarative_schema_changer = 'off';` or `SET use_declarative_schema_changer = 'unsafe_always';` instead of `SET use_declarative_schema_changer = $use_decl_sc;`


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138651)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138651#-OG6l0j50E0WUaxWvIXN:-OG79hXxD6B_aCLMgpr4:b-g2czik) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d976034f574b2c23e9a45523ef1c5b0421de34/pkg/sql/logictest/testdata/logic_test/row_level_security#L13)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

This is nice, I didn't know we could set variables like this in our tests.  

But what is the advantage to use this variable?
Wouldn't it be clearer to always use `SET use_declarative_schema_changer = 'on';` , `SET use_declarative_schema_changer = 'off';` or `SET use_declarative_schema_changer = 'unsafe_always';` instead of `SET use_declarative_schema_changer = $use_decl_sc;`
</blockquote></details>

I do this to preserve the original value of the variable, ensuring no assumptions need to be made about its state.


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (approved):
Looks good! Just one minor comment.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138651#-:-OGGK7p4EubM8YU_0Dy4:b1f594r)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @spilchen)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_policy.go` line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/138651#-OGGKIDK1X_MAR9Q2iTo:-OGGKIDLAnChKg4Sjh9N:b-4jy0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/37c1b784c89030dfd783e98c0806561b7e6c1f02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_policy.go#L15)):*
> ```Go
> )
> 
> func init() {
> ```

Just noting down here from our call to add a dep rule for add policy.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Implement CREATE/DROP POLICY in the DSC

Prior to this commit, attempts to create or drop a policy succeeded but
did not write any metadata to the table descriptor. This commit
introduces the first step in persisting policy metadata by storing the
policy name. Future PRs will handle storing additional policy details.

The policy name is now stored in a new field within the table
descriptor, and each policy is assigned a unique ID specific to the
table. This commit also adds elements to the DSC to support adding and
removing policy information in the descriptor. Support in the legacy
schema changer has been intentionally left out.

Epic: CRDB-11724
Informs #136696
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138651)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @michae2)

___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_policy.go` line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/138651#-OGGKIDK1X_MAR9Q2iTo:-OGGgV1k1WFMNJYy9Vcy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37c1b784c89030dfd783e98c0806561b7e6c1f02/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_policy.go#L15)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Just noting down here from our call to add a dep rule for add policy.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice! keeping the initial boilerplate as its own large PR was a good approach which should make future reviews pretty smooth.

> Reviewed 36 of 59 files at r1, 2 of 4 files at r2, 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138651#-:-OGHBBIu6DR0EEHEDx91:bcprsfq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Dedej-Bergin and @michae2)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR! 

bors r+
